### PR TITLE
Add TypeScript definition

### PR DESCRIPTION
**Pull request by Codex-**
Removed my request from DefinitelyTyped and migrated following your [guide.](https://github.com/sindresorhus/typescript-definition-style-guide)

Hopefully this meets the requirements :)

### CODE REVIEW

File: index.d.ts [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+
+/**
+ * Hooks stderr or alternatively the streams in options.
+ * 
+ * @param opts
+ * @param transform - the transform function.
+ * @returns a Promise with a unhook() method which, when called, unhooks
+ * the streams and resolves the Promise.
+ */
+export function stderr(opts: Options, transform: Transform): PromiseUnhook;
```
Comment (sindresorhus) [inline index.d.ts]: `opts` => `options`

File: index.d.ts [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+ * @returns a Promise with a unhook() method which, when called, unhooks
+ * the streams and resolves the Promise.
+ */
+export function stderr(transform: Transform): PromiseUnhook;
+
+/**
+ * Hooks stderr or alternatively the streams in options.
+ * 
+ * @param opts
+ * @param transform - the transform function.
```
Comment (sindresorhus) [inline index.d.ts]: Sentences should start with an uppercase character.

File: index.d.ts [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+	 * @default [process.stdout, process.stderr]
+	 */
+	streams?: Writable[];
+}
+
+/**
+ * Promise with a unhook() method which, when called, resolves the Promise
+ * with an empty result.
+ */
+export interface PromiseUnhook extends Promise<any> {
```
Comment (sindresorhus) [inline index.d.ts]: I would rather call this `HookPromise` as that's what it is. It's a Promise for the hook.

File: index.d.ts [hunk 4]
```diff
@@ -0,0 +1,92 @@
+// Type definitions for hook-std 1.1
+// Project: https://github.com/sindresorhus/hook-std
+// Definitions by: Alex Miller <https://github.com/codex->
+
+import {Writable} from 'stream';
+
+/**
+ * unhook() method which, when called, unhooks from a stream
```
Comment (sindresorhus) [inline index.d.ts]: unhook() should be wrapped in backticks.

File: index.d.ts [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+ * and resolves the Promise.
+ */
+export type Unhook = () => void;
+
+/**
+ * Receives stream output as the first argument and the unhook method as the
+ * second argument.
+ *
+ * Optionally, when in silent mode, you may return a boolean to influence
+ * the return value of .write(...).
```
Comment (sindresorhus) [inline index.d.ts]: .write(...) should be wrapped in backticks.

File: index.d.ts [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+ * Hooks stdout.
+ * 
+ * @param transform - the transform function.
+ * @returns a Promise with a unhook() method which, when called, unhooks
+ * the streams and resolves the Promise.
+ */
+export function stdout(transform: Transform): PromiseUnhook;
+
+/**
+ * Hooks stdout or alternatively the streams in options.
```
Comment (sindresorhus) [inline index.d.ts]: > alternatively the streams in options.

This is not correct. You're not supposed to use the `streams` option with the `.stdout`/`stderr` methods. I guess that needs to be made clearer in the docs.

File: index.d.ts [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+ * 
+ * @param opts
+ * @param transform - the transform function.
```
Comment (sindresorhus) [inline index.d.ts]: > If the parameter description just repeats the parameter name, leave it out.

**Review by sindresorhus — CHANGES_REQUESTED** [review]

File: index.d.ts [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+
+/**
+ * unhook() method which, when called, unhooks from a stream
+ * and resolves the Promise.
+ */
+export type Unhook = () => void;
+
+/**
+ * Receives stream output as the first argument and the unhook method as the
+ * second argument.
```
Comment (sindresorhus) [inline index.d.ts]: Don't hard-wrap. Applies to all the comments.

File: index.d.ts (hunk 3 above)
Comment (sindresorhus) [reply index.d.ts]: `any` => `void`?

File: index.d.ts (hunk 4 above)
Comment (Codex-) [reply index.d.ts]: Ah I see, wasn't sure if this was supported, cheers

**Comment by sindresorhus** [comment]
Let me know if there's anything in the style guide that could be clearer or improved. It's still a work-in-progress :)

File: index.d.ts (hunk 6 above)
Comment (Codex-) [reply index.d.ts]: Hmmm I suppose that means I should remove these:
`stderr(options: Options, transform: Transform): HookPromise;`
`stdout(options: Options, transform: Transform): HookPromise;`

Would you prefer a second set of options be typed in such a way that doesn't allow options with streams as parameter types for `stdout` and `stderr`?

I misunderstood originally but I think I understand that if you use hookstd via `var hookStd = require('hook-std')` then the call `hookStd(opts, str => {})` would be what consumes the stream options, so will need to rework some of the typings to suit this

**Comment by Codex-** [comment]
Re: the guide. I suppose the main thing I found a little tricky at first was how to use the type assertions correctly, I ended up going through a few other projects to understand the usage more correctly.

File: index.d.ts (hunk 6 above)
Comment (sindresorhus) [reply index.d.ts]: You could do something like this for just the default export:

```ts
type DefaultOptions = Options & {
	streams?: Writable[];
}

export default function hookStd(options?: DefaultOptions, transform: Transform): PromiseUnhook;
```

And remove `streams` from the `Options` interface

**Comment by sindresorhus** [comment]
Also note:

> For packages with a default export, use `export default foo;`, not `export = foo;`. You will have to add `module.exports.default = foo;` to the `index.js` file.

**Comment by sindresorhus** [comment]
You also left out types for the default export.

**Comment by Codex-** [comment]
Yup I've come to realize this after the comment around the stream options, will sort out soonish :)

File: index.d.ts (hunk 6 above)
Comment (Codex-) [reply index.d.ts]: I created a `DefaultOptions` interface which extends the `Options` interface to create this typing.

Thanks for the suggestion, have implemented similar, however you can't follow an optional param with a required one.

**Comment by Codex-** [comment]
I've _hopefully_ this time covered everything :)

**Comment by Codex-** [comment]
Mistakes were made...

**Comment by Codex-** [comment]
Should be good now.

File: index.d.ts (hunk 7 above)
Comment (Codex-) [reply index.d.ts]: Can you please clarify if this means to remove the whole `@param` doc or just the description?

**Comment by Codex-** [comment]
Actually I take it back, few issues to look at still, will comment when ready for another review

**Comment by Codex-** [comment]
Updated with the default export and fixed some typings. I think this is more up to scratch now. Sorry for the delays, uni has been crazy busy.

Can you please clarify a couple of things:

`If the parameter description just repeats the parameter name, leave it out.`
Does this mean I should remove all docstring params that are simply `@param transform` or that there shouldn't be a description if the description just repeats the param such as `@param transform the transform function`.


In the [readme](https://github.com/sindresorhus/hook-std#hookstdoptions-transform) it specifies `unhooks both stdout and stderr` but if you use custom or additional streams it unhooks these too, should it in that case read something more along the lines of `unhooks both stdout, stderr or any specified streams`

Cheers

**Comment by sindresorhus** [comment]
> Does this mean I should remove all docstring params that are simply @param transform

Yes

> In the readme it specifies unhooks both stdout and stderr but if you use custom or additional streams it unhooks these too, should it in that case read something more along the lines of unhooks both stdout, stderr or any specified streams

Maybe just:

> Unhooks the streams in the `streams` option.

File: index.d.ts [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+ * @returns A `Promise` with a `unhook()` method which, when called, unhooks the streams and resolves the `Promise`.
+ */
+export function stdout(transform: Transform): HookPromise;
+
+/**
+ * Hooks stdout.
+ *
+ * @returns A `Promise` with a `unhook()` method which, when called, unhooks the streams and resolves the `Promise`.
+ */
+export function stdout(options: Options, transform: Transform): HookPromise;
```
Comment (sindresorhus) [inline index.d.ts:66]: Why make this an overload instead of just marking `options` as optional?

Comment (Codex-) [reply index.d.ts:66]: Optional parameters cannot precede required parameters.

Comment (sindresorhus) [reply index.d.ts:66]: Hmm, ok. That's a lame limitation of TS...

**Comment by Codex-** [comment]
Just wondering if there were any pending changes you were wanting that I have possibly missed?

**Review by sindresorhus — APPROVED** [review]

**Comment by sindresorhus** [comment]
Thank you :)

**Comment by Codex-** [comment]
Great, thanks Sindre!

**Pull request merged**